packages/x-adapter/src/mappers/mapper.factory.ts

```diff
@@ -0,0 +1,111 @@
+import { deepMerge } from '@empathyco/x-deep-merge';
+import { isFunction, isObject, reduce, isPath, isArray } from '@empathyco/x-utils';
+import { Schema, SubSchemaTransformer } from '../schemas/schemas.types';
+import { Mapper, MapperContext } from '../types/mapper.types';
+import { extractValue } from '../utils/extract-value';
+
+/**
+ * The 'mapperFactory' function creates a {@link Mapper | mapper function} for a given
+ * {@link Schema | schema}.
+ *
+ * @param schema - The {@link Schema | schema} to apply in the {@link Mapper | mapper function}.
+ *
+ * @returns A {@link Mapper | mapper function} that applies the given {@link Schema | schema}.
+ *
+ * @public
+ */
+export function mapperFactory<Source, Target>(
+  schema: Schema<Source, Target>
+): Mapper<Source, Target | undefined> {
+  return function mapper(source: Source, context: MapperContext): Target | undefined {
+    return mapSchema(source, schema, context);
+  };
+}
+
+/**
+ * The `mapSchema()` function creates a new object populated with the transformations defined by a
+ * {@link Schema} applied to a source object.
+ *
+ * @param source - The object to apply the transformations to.
+ * @param schema - The object that defines the transformations to apply.
+ * @param context - The {@link MapperContext | mapper context} to feed the transformations with.
+ *
+ * @returns A new object with each element being the result of the applied transformation.
+ *
+ * @internal
+ */
+function mapSchema<Source, Target>(
+  source: Source,
+  schema: Schema<Source, Target>,
+  context: MapperContext
+): Target | undefined {
+  if (!source) {
+    return undefined;
+  }
+  return reduce(
+    schema,
+    (target, key, transformer) => {
+      if (typeof transformer === 'string' && isPath(source, transformer)) {
+        target[key] = extractValue(source, transformer) as Target[typeof key];
+      } else if (isFunction(transformer)) {
+        target[key] = transformer(source, context);
+      } else if (isObject(transformer)) {
+        if ('$subSchema' in transformer) {
+          const value = applySubSchemaTransformer<Source, Target[typeof key]>(
+            source,
+            transformer as SubSchemaTransformer<Source, Target[typeof key]>,
+            context,
+            schema as unknown as Schema<Source, Target[typeof key]>
+          ) as Target[keyof Target];
+
+          if (value) {
+            target[key] = value;
+          }
+        } else {
+          const value = mapSchema<Source, Target[typeof key]>(source, transformer, context);
+
+          if (value) {
+            target[key] = value;
+          }
+        }
```

Comment:
Can we extract `value` and get rid of the nested `if`s?

```suggestion
       let value: Target[keyof Target] | undefined;

        if ('$subSchema' in transformer) {
          value = applySubSchemaTransformer<Source, Target[typeof key]>(
            source,
            transformer as SubSchemaTransformer<Source, Target[typeof key]>,
            context,
            schema as unknown as Schema<Source, Target[typeof key]>
          ) as Target[keyof Target];
        } else {
          value = mapSchema<Source, Target[typeof key]>(source, transformer, context);
        }

        if (value) {
          target[key] = value;
        }
```

Reply:
Yes, we should. Thanks for the suggestion!

Reply:
Just realized this can be converted to a ternary. We might lose a little bit of code clarity so we can also leave it as is